### PR TITLE
fix(ssr): omit dangerous stylesheet hrefs and stop dropping valid CDN links

### LINKED ISSUE #74: fix(ssr): renderHead drops valid CDN stylesheets via broken origin allowlist
### axiom-framework version
0.9.9

### Runtime
Bun

### Runtime version
Bun 1.3.14

### Description
`renderHead()` (SSR) silently DROPS valid external CDN stylesheets. The stylesheet `href` was validated against a placeholder origin allowlist (`ALLOWED_STYLESHEET_ORIGIN`) that did not match real HTTPS CDN hosts, so legitimate `<link rel="stylesheet">` tags (e.g. `https://fonts.googleapis.com/...`) were never emitted. This also left SSR stylesheet sanitization INCONSISTENT with the attribute sanitizer in `src/core/attrs.ts` (which uses scheme-based deny-list semantics).

Expected: valid relative / http / https hrefs render exactly (HTML-escaped); only dangerous schemes (`javascript:`, `data:`, `vbscript:`, `file:`) and protocol-relative `//host` hrefs are blocked — and blocked ones are OMITTED (no `<link>` emitted), not rewritten.

### Minimal reproduction
```typescript
import { renderToString } from 'axiom-framework'

const html = renderToString(App, {
  metadata: { stylesheets: ['https://fonts.googleapis.com/css?family=Roboto'] }
})
// Bug: the <link rel="stylesheet"> for the Google Fonts CDN is MISSING from <head>
```

### Which phase is affected?
- [x] SSR (`src/ssr.ts` `renderHead`) + URL sanitization (`src/core/attrs.ts`)

### Additional context
Security hardening item from the C2 backlog. Fix replaces the broken origin allowlist with scheme-floor sanitization shared with `attrs.ts` (`sanitizeUrlValue`), and documents CSP `style-src` as the origin-enforcement layer in `SECURITY.md`. Part of #30 (v1.0.0 Readiness).

### PR DESCRIPTION
**Pull request by naml14**
## Summary

Fixes SSR `renderHead()` silently dropping valid external CDN stylesheets. The broken `ALLOWED_STYLESHEET_ORIGIN` placeholder allowlist is replaced with scheme-floor sanitization shared with `src/core/attrs.ts` (`sanitizeUrlValue`). Dangerous schemes (`javascript:`, `data:`, `vbscript:`, `file:`) and protocol-relative `//host` hrefs are now OMITTED (no `<link>` emitted), while valid relative/http/https hrefs render exactly and HTML-escaped.

> Chained PR: base is `feat/starter-demo-onboarding` (this fix depends on the `attrs.ts` sanitizer that only exists on that branch). Merge after the base PR.

## Type of change

- [x] `fix` — bug fix (patch bump)

## Checklist

- [x] Tests added or updated for the changed behavior (20 tests: 10 unit + 10 SSR)
- [x] `bun test` passes locally (676 pass / 2 skip / 0 fail)
- [x] `bun run typecheck` passes locally (clean)
- [x] This PR links at least one issue (`Closes #74`)
- [ ] `docs/V1-0-0-EVIDENCE-LOG.md` was updated — N/A (file does not exist in this branch)
- [x] Commit messages follow Conventional Commits format
- [x] Hot path invariant respected: no DOM reads added to reflow/fast-path/flex/commit (change is SSR-only)
- [x] No runtime dependencies added to `src/`

## Related issues

Closes #74
Part of #30

## Testing notes

- `bun test tests/ssr.test.ts tests/edge-cases.test.ts` → 85 pass / 0 fail.
- Public `sanitizeAttrValue()` contract is UNCHANGED (component attrs still return `'#blocked'`); `sanitizeUrlValue` is module-internal and NOT re-exported from `src/index.ts`.
- `dist/` is gitignored and regenerated by `prepublishOnly` (typecheck && test && build), so no stale artifact ships.
- Verify report: `openspec/changes/ssr-stylesheet-origin-policy/verify-report.md` (PASS_WITH_WARNINGS).

## Summary by Sourcery

Actualizar el manejo de las hojas de estilo en SSR para compartir la sanitización de URLs con los atributos del núcleo, restaurando el renderizado válido de hojas de estilo desde CDN mientras se omiten href peligrosos y se documenta el modelo de seguridad.

Correcciones de errores:
- Evitar que `SSR renderHead()` descarte silenciosamente URLs válidas de hojas de estilo externas HTTPS y HTTP proporcionadas vía `metadata.stylesheets`.
- Asegurar que href de hojas de estilo peligrosos que usan `javascript:`, `data:`, `vbscript:`, `file:` o URLs con protocolo relativo no se emitan en la salida SSR.

Mejoras:
- Introducir un helper compartido `sanitizeUrlValue` en el módulo de atributos del núcleo para centralizar la validación de esquemas de URL en contextos sensibles a URLs.

Documentación:
- Revisar `SECURITY.md` para describir la nueva protección de “scheme-floor” para hojas de estilo y recomendar `CSP style-src` más validación del host del consumidor para la aplicación de origen.

Pruebas:
- Agregar pruebas unitarias enfocadas para `sanitizeUrlValue` y pruebas de SSR que cubran casos de metadatos de hojas de estilo seguros, peligrosos, mixtos y vacíos.
- Agregar artefactos de especificación, diseño, exploración, tareas y verificación de openspec para el cambio de política de origen de hojas de estilo en SSR.

Tareas de mantenimiento:
- Agregar y actualizar artefactos de cambio de openspec para rastrear el diseño, la implementación y la verificación de la política de origen de hojas de estilo en SSR.

<details>
<summary>Original summary in English</summary>

## Summary by Sourcery

Update SSR stylesheet handling to share URL sanitization with core attrs, restoring valid CDN stylesheet rendering while omitting dangerous hrefs and documenting the security model.

Bug Fixes:
- Prevent SSR renderHead() from silently dropping valid external HTTPS and HTTP stylesheet URLs provided via metadata.stylesheets.
- Ensure dangerous stylesheet hrefs using javascript:, data:, vbscript:, file:, or protocol-relative URLs are not emitted into SSR output.

Enhancements:
- Introduce a shared sanitizeUrlValue helper in the core attrs module to centralize URL scheme validation for URL-sensitive contexts.

Documentation:
- Revise SECURITY.md to describe the new stylesheet scheme-floor protection and recommend CSP style-src plus consumer host validation for origin enforcement.

Tests:
- Add focused unit tests for sanitizeUrlValue and SSR tests covering safe, dangerous, mixed, and empty stylesheet metadata cases.
- Add openspec specification, design, exploration, tasks, and verification artifacts for the SSR stylesheet origin policy change.

Chores:
- Add and update openspec change artifacts to track design, implementation, and verification of the SSR stylesheet origin policy.

</details>

### CODE REVIEW

File: src/core/attrs.ts [hunk 1]
```diff
@@ -206,6 +206,35 @@ export function isValidAttrName(key: string): boolean {
...
+    return '#blocked'
+  }
+  if (SAFE_URL_PATTERN.test(value)) {
+    return value
+  }
+  if (DANGEROUS_URL_SCHEME_RE.test(value)) {
+    return '#blocked'
+  }
+  return value
+}
```
[the remark this answers is not in the document]
Comment (naml14) [reply src/core/attrs.ts:242]: ✅ Resuelto. Se agregó la comprobación defensiva 	ypeof value !== 'string' al inicio de sanitizeUrlValue(). Si el valor no es string, retorna '#blocked' inmediatamente. Tests pasan (101 tests en archivos relevantes, 692 total).

File: src/ssr.ts [hunk 2]
```diff
@@ -198,7 +198,9 @@ function renderHead(metadata?: SSRMetadata): string {
 
   if (metadata.stylesheets !== undefined) {
     for (const href of metadata.stylesheets) {
-      html += `<link rel="stylesheet" href="${escapeHtml(href)}">`
+      const sanitizedHref = sanitizeUrlValue(href)
+      if (sanitizedHref === '#blocked') continue
+      html += `<link rel="stylesheet" href="${escapeHtml(sanitizedHref)}">`
     }
```
[the remark this answers is not in the document]
Comment (naml14) [reply src/ssr.ts:211]: ✅ Resuelto. Se agregaron las comprobaciones defensivas en 
enderHead():\n- 	ypeof href !== 'string': salta valores no-string\n- href.trim() === '': salta strings vacíos o solo espacios\n\nEsto evita excepciones de runtime y la emisión de <link> tags con href vacío. Tests pasan (101 tests en archivos relevantes, 692 total).

File: tests/edge-cases.test.ts [hunk 3]
```diff
@@ -432,6 +433,50 @@ describe('security: attribute sanitization module', () => {
       })
     })
   })
+
+  describe('sanitizeUrlValue', () => {
```
[the remark this answers is not in the document]
Comment (naml14) [reply tests/edge-cases.test.ts:460]: ✅ Resuelto. Se agregaron tests para los patrones seguros documentados:\n- mailto:user@example.com\n- 	el:+1-555-123-4567\n- #section / #\n- custom:something / pp://resource\n\nTodos los casos seguros pasan y los peligrosos siguen bloqueados. Tests pasan (101 tests en archivos relevantes, 692 total).

File: openspec/changes/ssr-stylesheet-origin-policy/verify-report.md [hunk 4]
```diff
@@ -0,0 +1,300 @@
...
+| Stylesheet Scheme Validation | Block javascript: URI scheme | `tests/ssr.test.ts > javascript: href is omitted entirely — no <link> tag emitted` | ✅ COMPLIANT |
+| Stylesheet Scheme Validation | Block data: URI scheme | `tests/ssr.test.ts > data: href is omitted entirely` | ✅ COMPLIANT |
+| Stylesheet Scheme Validation | Block vbscript: and file: URI schemes | `tests/ssr.test.ts > vbscript: and file: hrefs are omitted entirely` | ✅ COMPLIANT |
+| Stylesheet Scheme Validation | Block protocol-relative URLs | `tests/ssr.test.ts > protocol-relative href is omitted entirely` | ✅ COMPLIANT |
+| Stylesheet Scheme Validation | HTML escape valid hrefs | `tests/ssr.test.ts > href with HTML-special chars is escaped in output` | ✅ COMPLIANT |
+| Stylesheet Scheme Validation | Render multiple mixed stylesheets | `tests/ssr.test.ts > mixed list: safe hrefs emitted, dangerous ones omitted` | ✅ COMPLIANT |
+| Stylesheet Scheme Validation | Handle empty or undefined stylesheets | `tests/ssr.test.ts > empty stylesheets array emits no link tags` and `undefined stylesheets emits no link tags` | ✅ COMPLIANT |
```
[the remark this answers is not in the document]
Comment (naml14) [reply openspec/changes/ssr-stylesheet-origin-policy/verify-report.md:205]: ✅ Resuelto. Se corrigio la concordancia sujeto-verbo: 'undefined stylesheets emit no link tags'. Tests pasan.

File: openspec/changes/ssr-stylesheet-origin-policy/proposal.md [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+Furthermore, while the framework provides robust URL-scheme sanitization for component attributes in `src/core/attrs.ts` (blocking `javascript:`, `data:`, `vbscript:`, `file:`, and protocol-relative `//` URLs), this logic is NOT applied to stylesheet `href`s in `renderHead()`. This creates a security inconsistency where dangerous schemes might be emitted in the document `<head>`.
+
+**Why now?**
+The current implementation breaks the public `metadata.stylesheets` API contract for anyone using an external CDN. We need to fix this bug immediately while ensuring the framework maintains a consistent security posture against provably-dangerous URI schemes.
+
+## 2. Goals and Non-Goals
+
+### Goals
+- **Fix the bug**: Remove the hardcoded placeholder regex to restore the public API contract for `metadata.stylesheets`, allowing legitimate absolute URLs.
+- **Enforce Scheme Consistency**: Apply the existing scheme-blocking logic from `attrs.ts` to stylesheet `href`s in `renderHead()`, ensuring dangerous schemes are blocked and emitted as `#blocked`.
```
[the remark this answers is not in the document]
Comment (naml14) [reply openspec/changes/ssr-stylesheet-origin-policy/proposal.md]: Fixed. proposal.md Success Criteria #2 now correctly states 'omitted entirely' (no link tag) instead of '#blocked'. Ver: d7180d1.

File: openspec/changes/ssr-stylesheet-origin-policy/proposal.md [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+
+### Out of Scope
+- Adding new configuration options to `SSRRenderOptions`.
+- Changing how other metadata fields (e.g., `scripts`, `links`) are handled, unless directly tied to the extracted scheme-blocking logic.
+- Implementing CSP generation within the framework itself.
+
+## 4. High-Level Approach
+
+1. **Extract/Reuse Scheme Blocker**: In `src/core/attrs.ts`, ensure the logic that blocks `javascript:`, `data:`, `vbscript:`, `file:`, and protocol-relative `//` is accessible (e.g., as a `sanitizeHref` utility).
+2. **Fix `renderHead()`**: Remove the `ALLOWED_STYLESHEET_ORIGIN` regex in `src/ssr.ts`. Iterate over `metadata.stylesheets` and pass each `href` through the scheme blocker before emitting the `<link rel="stylesheet">` tag. If blocked, emit `#blocked` (consistent with component attrs behavior).
```
[the remark this answers is not in the document]
Comment (naml14) [reply openspec/changes/ssr-stylesheet-origin-policy/proposal.md]: Fixed. proposal.md High-Level Approach step 2 now correctly states 'omit the tag entirely' instead of 'emit #blocked'. Ver: d7180d1.

File: openspec/changes/ssr-stylesheet-origin-policy/proposal.md [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+
+- **Risk: Breaking changes to consumers.**
+  - **Mitigation**: The current behavior is already broken (silently dropping valid CDNs). This change fixes the break. By explicitly blocking only provably-dangerous schemes (which are never valid for stylesheets), the false positive rate is ~0.
+- **Risk: Incomplete protection theater.**
+  - **Mitigation**: We explicitly document in `SECURITY.md` that scheme-blocking does NOT prevent CSS exfiltration to attacker-controlled `https://` domains. We clearly delegate origin validation to the consumer and CSP, setting accurate security expectations.
+
+## 6. Success Criteria
+
+1. Real CDN URLs (e.g., `https://fonts.googleapis.com/...`) and relative URLs correctly render in the SSR output without being dropped.
+2. Dangerous schemes (`javascript:`, `data:`, `vbscript:`, `file:`, `//`) are blocked and replaced with `#blocked` in the output.
```
[the remark this answers is not in the document]
Comment (naml14) [reply openspec/changes/ssr-stylesheet-origin-policy/proposal.md]: Fixed. proposal.md now correctly states 'omitted entirely' instead of 'replaced with #blocked' in Success Criteria #2. Ver: d7180d1.

File: openspec/changes/ssr-stylesheet-origin-policy/tasks.md [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+
+| Field | Value |
+|-------|-------|
+| Estimated changed lines | 140-220 |
+| Estimated implementation files | 6 |
+| 400-line budget risk | Low |
+| Chained PRs recommended | No |
+| Suggested split | single PR |
+| Delivery strategy | auto-forecast |
+| Chain strategy | pending |
```
[the remark this answers is not in the document]
Comment (naml14) [reply openspec/changes/ssr-stylesheet-origin-policy/tasks.md]: Fixed. tasks.md now correctly states 'Chained PRs recommended: Yes' with 'stacked-to-main' strategy, reflecting that this PR IS chained to feat/starter-demo-onboarding. Ver: d7180d1.

File: openspec/changes/ssr-stylesheet-origin-policy/tasks.md [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+
+Decision needed before apply: No
+Chained PRs recommended: No
+Chain strategy: pending
+400-line budget risk: Low
```
[the remark this answers is not in the document]
Comment (naml14) [reply openspec/changes/ssr-stylesheet-origin-policy/tasks.md:18]: Fixed. tasks.md summary block now correctly says 'Chained PRs recommended: Yes' with chain strategy 'stacked-to-main', consistent with the PR being chained. Ver: d7180d1.

File: SECURITY.md [hunk 10]
```diff
@@ -116,22 +116,32 @@ await renderToString(app, { metadata: { inlineStyles: safeCss } });
 
 ### External Stylesheet SSRF
```
[the remark this answers is not in the document]
Comment (naml14) [reply SECURITY.md]: Fixed. SECURITY.md section renamed from 'External Stylesheet SSRF' to 'External Stylesheet Origin Validation' for accurate threat model terminology (CSS exfiltration, not SSRF). Ver: d7180d1.
